ledger/src/blockstore_processor.rs

```diff
@@ -1574,6 +1582,26 @@ pub fn fill_blockstore_slot_with_ticks(
     last_entry_hash
 }
 
+/// Check the transaction execution results to see if any instruction errored by exceeding the
+/// accounts data budget.  If yes, the whole block needs to be failed.
+fn check_accounts_data_budget<'a>(
+    execution_results: impl IntoIterator<Item = &'a TransactionExecutionResult>,
+) -> Result<()> {
+    if execution_results.into_iter().any(|execution_result| {
+        matches!(
+            execution_result.flattened_result(),
+            Err(TransactionError::InstructionError(
+                _,
+                InstructionError::AccountsDataBudgetExceeded
```

Comment:
Hmm isn't there already a `TransactionError::WouldExceedMaxAccountCostLimit` for this case? Do we need a separate `InstructionError`?

Reply:
Yes, so the checking is currently happening _per instruction_ in InvokeContext. This helps guard against the case where a single _transaction_ may have the max instructions create the max sized new accounts; so the first instruction that exceeds the max would fail with the InstructionError. So a new InstructionError was added to handle that.

Then, by the time we get to replay, we have a TransactionError that wraps the InstructionError. So I need to check if there was any instruction that exceeded the max. This section of code highlighted is the `matches!()` part. The `Err()` part below could use `WouldExceedMaxAccountCostLimit` instead, yes. I replied to the other comment here: https://github.com/solana-labs/solana/pull/23422#discussion_r817190856

Reply:
Edit: Hmm I see so there's a different error for per instruction exceeding and block level exceeding. One implies the other no? Do we need to check for both?


1) Seems like we can just use the existing `TransactionError::WouldExceedMaxAccountDataCostLimit`
2) I would rename this variant of `TransactionError` to convey this difference, i.e. `WouldExceedMaxAccountDataCostLimit -> WouldExceedMaxBlockAccountData`

Reply:
Hah, yes, this is not straightforward... There are two different limits:
1. There is a limit on the amount the account data can grow _within a block_. This is set to 100 MB and is in master.
2. There is/will be a limit on the amount the account data can be _in total_. This is set to 128 GB.

This PR is to address (2). Prior work has been added to check if any instruction exceeds to total account data size. Now, I need to fail the whole block if any instruction returned that error during execution (`InstructionError::AccountsDataBudgetExceeded.`).

> Hmm I see so there's a different error for per instruction exceeding and block level exceeding. One implies the other no? Do we need to check for both?
We do need to check for both, as they are separate error cases. For example:
- Assume the total account data size is 40 GB. A block could have transactions that create a bunch of accounts, totaling 101 MB. In this case the total account data size would still be below the 128 limit, but the account data size limit _per block_ would've been exceeded.
- Assume now the total account data size is currently at 127,999,999,801 bytes (199 bytes short of 128 GB). A block has a single transaction to split a stake account into a new account (which has a size of 200 bytes). The new account data in this block is only 200 B, so well below 100 MB. However, now we've exceeded the total account data size.

I like your suggestion:
> The variant `TransactionError::InstructionError` could just wrap that existing `InstructionError::AccountsDataBudgetExceeded`
Or I can stick with the two different Transaction Errors.



Hopefully this thread will resolve the other comment here: https://github.com/solana-labs/solana/pull/23422#discussion_r817190856

@carllin Wdyt?

Reply:
Ok so:

1. AccountsDataBudgetExceeded - This is the whole accounts 128GB budget, so this should be renamed to `MaxAccountsDataSizeExceeded` for clarity, with a comment over the enum variant (your description above was helpful thanks!) of what this error implies.

As for the return value here, I think we can just wrap the `InstructionError` as I suggested earlier, no need to introduce more complexity. This is also consistent with any other `InstructionError`, as they don't have another `TransactionError` either

2. The existing `WouldExceedMaxAccountDataCostLimit` is the intra-block limit? Seems like this should be renamed to `BlockAccountsDataSizeExceeded` (lets keep the naming convention consistent with 1.), and add a comment describing what this implies (taken from your comment above as well)

Reply:
I've renamed `AccountsDataBudgetExceeded ` to `MaxAccountsDataSizeExceeded`. This does touch a lot of existing code already using this error. Will this be an issue?